feat: context middleware for insights headers

lzap opened the pull request:
Adds new middleware that stores custom HTTP headers into context and callback function that puts the value into all log records when present. Improved callback mechanism, added tests for callback mechanism and extended the example to see it in action.

This will be only useful for image-builder service (insights service) where `X-RH-Request-Id` and `Edge-Id` are sent so logs can be more easily correlated with requests from 3Scale gateway. This is completely optional feature.

lzap commented on go.work:
```diff
@@ -4,11 +4,11 @@ toolchain go1.23.4
 	./internal/example_splunk
 	./internal/example_web
-	./internal/example_cli
 )
```
Happens when you do `go work sync`

lzap commented on internal/example_web/main.go:
```diff
@@ -27,30 +27,36 @@ func subProcess(ctx context.Context) {
 		WithResponseHeader: true,
 		Filters:            []strc.Filter{strc.IgnorePathPrefix("/metrics")},
 	})
+	m3 := strc.HeadfieldPairMiddleware(pairs)
```
Example how to use this middleware - the incoming request ID header will be stored in context and then extracted and added as a log field to every single log message that is logged with the context.

lzap commented on pkg/logrus/proxy.go:
```diff
@@ -92,35 +94,19 @@ func (p *Proxy) Trace(args ...any) {
-		p.dest.ErrorContext(p.ctx, strings.Join(anyToString(args), " "))
-	}
+	p.dest.ErrorContext(p.ctx, strings.Join(anyToString(args), " "))
 }
```
Another bugs found by the new tests.

lzap commented:
When working on edge service, I ran into more logrus methods being used so added bunch of missing methods and interfaces. None of this is used in image builder but this makes the logrus proxy slightly more complete. Again, this is only a temporary API to ease the transition process.

schuellerf commented on pkg/strc/middleware_ctx.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+// HeadfieldPairMiddleware is a middleware that extracts values from headers and stores them in the context.
+func HeadfieldPairMiddleware(pairs []HeadfieldPair) func(http.Handler) http.Handler {
```
Is there any test for those?

schuellerf replied:
(and should there be some? ;) )

lzap replied:
Added bunch of tests.

lzap commented:
Rebased, ready for re-review. While working on `edge` service, I had to add many more `logrus` messages, I improved initialization of `logrus` and `strc` default destinations to be thread-safe, added more documentation for middleware, added bunch of tests for all `strc` middlewares and found out that filtering was not working as expected so fixed it. Let me comment on those places in the code.

lzap commented on internal/example_logrus/main.go:
```diff
@@ -35,19 +40,22 @@ func main() {
+	logrus.WithError(errors.New("test")).Error("msg with err")
 
 	logrus.WithFields(logrus.Fields{
 		"str":   "value",
```
Some functions were not proxied.

lzap commented on pkg/logrus/interface.go:
```diff
@@ -0,0 +1,199 @@
+	Tracef(format string, args ...interface{})
+	Trace(args ...interface{})
+	Traceln(args ...interface{})
+}
```
I copy-pasted the intefaces directly from the `logrus` project this time to ensure 100% compatibility of the interface.

lzap commented on pkg/logrus/proxy.go:
```diff
@@ -3,22 +3,10 @@ package logrus
-	DebugLevel
-	TraceLevel
+	"sync/atomic"
 )
```
Types and constants moved to `interface.go`. This now uses `atomic.Pointer` for thread-safe initialization.

lzap commented on pkg/logrus/proxy.go:
```diff
@@ -196,113 +218,133 @@ func (p *Proxy) Panicln(args ...any) {
 func Panicln(args ...any) {
-	proxy.Panicln(args...)
+	proxy.Load().Panicln(args...)
 }
```
I did not encounter any thread-safety issues, but since this is a good practice to do and `slog` library itself uses the same technique, why not to use it.

lzap commented on pkg/strc/middleware.go:
```diff
@@ -93,6 +114,14 @@ func NewMiddlewareWithConfig(logger *slog.Logger, config MiddlewareConfig) func(
+					return
+				}
+			}
+
```
Moved filters to exit earlier, so paths like `/metrics` do not have any trace/span processing at all, they do not need to.

lzap commented on pkg/strc/middleware_recover.go:
```diff
@@ -0,0 +1,34 @@
+			next.ServeHTTP(w, r)
+		})
+	}
+}
```
This file was renamed, now it returns a function so logger can be passed in for easier testing.

lzap commented on pkg/strc/multi.go:
```diff
@@ -134,8 +138,9 @@ func (h *MultiHandler) WithGroup(name string) slog.Handler {
+		inSpanGroup: name == SpanGroupName,
 	}
 }
 
```
All these changes improves and simplifies the callback mechanism which is used to inject context values from HTTP headers.

lzap commented on pkg/strc/trace.go:
```diff
@@ -25,14 +26,22 @@ var ParentIDName string = "parent"
+	return destination.Load()
 }
 
 type Span struct {
```
Similar change as in `logrus` for initialization thread safety.

schuellerf reviewed (approved):
I'll do some more examination and tests but it LGTM

lzap commented:
> I'll do some more examination and tests but it LGTM

Thanks, a bit confusing. Does this mean we can merge? :-)

lzap commented:
Thank you, merging then.
